src/Components/Components/test/Rendering/RenderTreeBuilderTest.cs

```diff
@@ -1831,6 +1831,69 @@ public void ProcessDuplicateAttributes_CanRemoveOverwrittenAttributes()
                 f => AssertFrame.Attribute(f, "3", "see ya"));
         }
 
+        [Fact]
+        public void AcceptsClosedFramesAsValid()
+        {
+            // Arrange
+            var builder = new RenderTreeBuilder();
+            var component = new TestComponent();
+            builder.OpenElement(0, "myElem");
+            builder.OpenRegion(1);
+            builder.OpenComponent<OtherComponent>(2);
+            builder.CloseComponent();
+            builder.CloseRegion();
+            builder.CloseElement();
```

Comment:
Do we have validation for when the close statements balance out but are in different order? For example
```
builder.OpenElement()
builder.OpenRegion().
builder.CloseElement();
builder.CloseRegion()
```

Reply:
It's a good point to raise, but I think we're OK with this one. In practice all of the `Close` methods will close whatever frame is outstanding on the stack, so even if you do deliberately get the order wrong it won't be harmful. The only weirdness you could cause by this would be failing to pass parameters to a child component because that's the only difference between the close methods.

Considering how perf-critical is this low-level part of rendering, we wouldn't want penalize people using `.razor` to produce normal components by introduce extra checking logic for cases that a valid `.razor` component could never trigger. We've always been clear that anyone writing rendertreebuilder logic manually is in the same situation as people writing C# logic manually: if you write a bug, you have a bug.